File: backend/pkg/core/worker/status.go
package worker

import (
	"sync"
	"time"
)

// WorkerState represents the current state of a worker
type WorkerState string

const (
	StateIdle      WorkerState = "idle"
	StateRunning   WorkerState = "running"
	StateCompleted WorkerState = "completed"
	StateFailed    WorkerState = "failed"
)

// WorkerStatus represents the current status of a worker
type WorkerStatus struct {
	Name        string      `json:"name"`
	State       WorkerState `json:"state"`
	Progress    int         `json:"progress"`     // 0-100
	Message     string      `json:"message"`      // Current operation description
	StartedAt   *time.Time  `json:"started_at"`   // When the current/last run started
	CompletedAt *time.Time  `json:"completed_at"` // When the last run completed
	Error       string      `json:"error"`        // Error message if failed
	ItemsTotal  int         `json:"items_total"`  // Total items to process
	ItemsDone   int         `json:"items_done"`   // Items processed so far
}

// StatusUpdate is sent to subscribers when status changes
type StatusUpdate struct {
	WorkerName string       `json:"worker_name"`
	Status     WorkerStatus `json:"status"`
	Timestamp  time.Time    `json:"timestamp"`
}

// StatusSubscriber is a channel that receives status updates
type StatusSubscriber chan StatusUpdate

// StatusService manages worker status tracking and broadcasting
type StatusService struct {
	statuses    map[string]*WorkerStatus
	subscribers map[StatusSubscriber]struct{}
	mu          sync.RWMutex
}

// Global status service instance
var (
	globalStatusService *StatusService
	statusOnce          sync.Once
)

// GetStatusService returns the global status service instance
func GetStatusService() *StatusService {
	statusOnce.Do(func() {
		globalStatusService = &StatusService{
			statuses:    make(map[string]*WorkerStatus),
			subscribers: make(map[StatusSubscriber]struct{}),
		}
	})
	return globalStatusService
}

// RegisterWorker initializes status tracking for a worker
func (s *StatusService) RegisterWorker(name string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if _, exists := s.statuses[name]; !exists {
		s.statuses[name] = &WorkerStatus{
			Name:  name,
			State: StateIdle,
		}
	}
}

// UpdateStatus updates a worker's status and broadcasts to subscribers
func (s *StatusService) UpdateStatus(name string, state WorkerState, progress int, message string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	status, exists := s.statuses[name]
	if !exists {
		status = &WorkerStatus{Name: name}
		s.statuses[name] = status
	}

	status.State = state
	status.Progress = progress
	status.Message = message

	if state == StateRunning && status.StartedAt == nil {
		now := time.Now()
		status.StartedAt = &now
		status.CompletedAt = nil
		status.Error = ""
	}

	if state == StateCompleted || state == StateFailed {
		now := time.Now()
		status.CompletedAt = &now
	}

	s.broadcast(name, *status)
}

// SetProgress updates progress with item counts
func (s *StatusService) SetProgress(name string, itemsDone, itemsTotal int, message string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	status, exists := s.statuses[name]
	if !exists {
		return
	}

	status.ItemsDone = itemsDone
	status.ItemsTotal = itemsTotal
	status.Message = message

	if itemsTotal > 0 {
		status.Progress = (itemsDone * 100) / itemsTotal
	}

	s.broadcast(name, *status)
}

// SetError sets an error message on a worker status
func (s *StatusService) SetError(name string, err string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	status, exists := s.statuses[name]
	if !exists {
		return
	}

	status.Error = err
	status.State = StateFailed
	now := time.Now()
	status.CompletedAt = &now

	s.broadcast(name, *status)
}

// StartWorker marks a worker as starting
func (s *StatusService) StartWorker(name string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	status, exists := s.statuses[name]
	if !exists {
		status = &WorkerStatus{Name: name}
		s.statuses[name] = status
	}

	now := time.Now()
	status.State = StateRunning
	status.StartedAt = &now
	status.CompletedAt = nil
	status.Progress = 0
	status.Message = "Starting..."
	status.Error = ""
	status.ItemsDone = 0
	status.ItemsTotal = 0

	s.broadcast(name, *status)
}

// CompleteWorker marks a worker as completed
func (s *StatusService) CompleteWorker(name string, message string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	status, exists := s.statuses[name]
	if !exists {
		return
	}

	now := time.Now()
	status.State = StateCompleted
	status.CompletedAt = &now
	status.Progress = 100
	status.Message = message

	s.broadcast(name, *status)
}

// ResetWorker resets a worker to idle state
func (s *StatusService) ResetWorker(name string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	status, exists := s.statuses[name]
	if !exists {
		return
	}

	status.State = StateIdle
	status.Progress = 0
	status.Message = ""
	status.ItemsDone = 0
	status.ItemsTotal = 0
	// Keep StartedAt and CompletedAt for "last run" info

	s.broadcast(name, *status)
}

// GetStatus returns the current status of a worker
func (s *StatusService) GetStatus(name string) *WorkerStatus {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if status, exists := s.statuses[name]; exists {
		// Return a copy
		statusCopy := *status
		return &statusCopy
	}
	return nil
}

// GetAllStatuses returns the current status of all workers
func (s *StatusService) GetAllStatuses() map[string]WorkerStatus {
	s.mu.RLock()
	defer s.mu.RUnlock()

	result := make(map[string]WorkerStatus, len(s.statuses))
	for name, status := range s.statuses {
		result[name] = *status
	}
	return result
}

// Subscribe adds a subscriber to receive status updates
func (s *StatusService) Subscribe() StatusSubscriber {
	s.mu.Lock()
	defer s.mu.Unlock()

	ch := make(StatusSubscriber, 100) // Buffered to prevent blocking
	s.subscribers[ch] = struct{}{}
	return ch
}

// Unsubscribe removes a subscriber
func (s *StatusService) Unsubscribe(ch StatusSubscriber) {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.subscribers, ch)
	close(ch)
}

// broadcast sends a status update to all subscribers (must be called with lock held)
func (s *StatusService) broadcast(name string, status WorkerStatus) {
	update := StatusUpdate{
		WorkerName: name,
		Status:     status,
		Timestamp:  time.Now(),
	}

	for ch := range s.subscribers {
		select {
		case ch <- update:
		default:
			// Channel full, skip this subscriber
		}
	}
}

// GetSubscriberCount returns the number of active subscribers
func (s *StatusService) GetSubscriberCount() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return len(s.subscribers)
}

File: backend/src/core/worker/status.rs
use std::{
    collections::BTreeMap,
    sync::{
        Arc, Mutex, MutexGuard,
        atomic::{AtomicU64, Ordering},
    },
};

use chrono::{DateTime, Utc};
use tokio::sync::mpsc;
use uuid::Uuid;

use super::{WorkerState, WorkerStatus, WorkerStatusUpdate};

const SUBSCRIBER_CAPACITY: usize = 100;

pub trait Clock: Send + Sync {
    fn now(&self) -> DateTime<Utc>;
}

#[derive(Debug, Default)]
pub struct SystemClock;

impl Clock for SystemClock {
    fn now(&self) -> DateTime<Utc> {
        Utc::now()
    }
}

type SendUpdate = dyn Fn(&WorkerStatusUpdate) -> SubscriberState + Send + Sync;

struct Subscriber {
    send: Box<SendUpdate>,
    is_closed: Box<dyn Fn() -> bool + Send + Sync>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum SubscriberState {
    Active,
    Closed,
}

#[derive(Default)]
struct StatusState {
    statuses: BTreeMap<String, WorkerStatus>,
    subscribers: BTreeMap<u64, Subscriber>,
}

pub struct StatusService {
    state: Mutex<StatusState>,
    clock: Arc<dyn Clock>,
    next_subscriber_id: AtomicU64,
}

impl StatusService {
    pub fn new(clock: Arc<dyn Clock>) -> Self {
        Self {
            state: Mutex::new(StatusState::default()),
            clock,
            next_subscriber_id: AtomicU64::new(1),
        }
    }

    pub fn register_worker(&self, name: &str) {
        let mut state = self.state();
        state
            .statuses
            .entry(name.to_owned())
            .or_insert_with(|| WorkerStatus::idle(name));
    }

    pub fn update_status(
        &self,
        name: &str,
        worker_state: WorkerState,
        progress: i32,
        message: impl Into<String>,
    ) {
        let now = self.clock.now();
        let mut state = self.state();
        let status = state
            .statuses
            .entry(name.to_owned())
            .or_insert_with(|| WorkerStatus::idle(name));
        status.state = worker_state;
        status.progress = progress;
        status.message = message.into();
        if worker_state == WorkerState::Running && status.started_at.is_none() {
            status.started_at = Some(now);
            status.completed_at = None;
            status.error.clear();
        }
        if matches!(worker_state, WorkerState::Completed | WorkerState::Failed) {
            status.completed_at = Some(now);
        }
        let status = status.clone();
        Self::broadcast(&mut state, now, name, status);
    }

    pub fn set_progress(
        &self,
        name: &str,
        items_done: i32,
        items_total: i32,
        message: impl Into<String>,
    ) {
        let now = self.clock.now();
        let mut state = self.state();
        let Some(status) = state.statuses.get_mut(name) else {
            return;
        };
        status.items_done = items_done;
        status.items_total = items_total;
        status.message = message.into();
        if items_total > 0 {
            status.progress = items_done.saturating_mul(100) / items_total;
        }
        let status = status.clone();
        Self::broadcast(&mut state, now, name, status);
    }

    pub fn set_error(&self, name: &str, error: impl Into<String>) {
        let now = self.clock.now();
        let mut state = self.state();
        let Some(status) = state.statuses.get_mut(name) else {
            return;
        };
        status.error = error.into();
        status.state = WorkerState::Failed;
        status.completed_at = Some(now);
        let status = status.clone();
        Self::broadcast(&mut state, now, name, status);
    }

    pub fn start_worker(&self, name: &str, task_run_id: Option<Uuid>) {
        let now = self.clock.now();
        let mut state = self.state();
        let status = state
            .statuses
            .entry(name.to_owned())
            .or_insert_with(|| WorkerStatus::idle(name));
        status.state = WorkerState::Running;
        status.task_run_id = task_run_id;
        status.started_at = Some(now);
        status.completed_at = None;
        status.progress = 0;
        status.message = "Starting...".to_owned();
        status.error.clear();
        status.items_done = 0;
        status.items_total = 0;
        let status = status.clone();
        Self::broadcast(&mut state, now, name, status);
    }

    pub fn complete_worker(&self, name: &str, message: impl Into<String>) {
        let now = self.clock.now();
        let mut state = self.state();
        let Some(status) = state.statuses.get_mut(name) else {
            return;
        };
        status.state = WorkerState::Completed;
        status.completed_at = Some(now);
        status.progress = 100;
        status.message = message.into();
        let status = status.clone();
        Self::broadcast(&mut state, now, name, status);
    }

    pub fn reset_worker(&self, name: &str) {
        let now = self.clock.now();
        let mut state = self.state();
        let Some(status) = state.statuses.get_mut(name) else {
            return;
        };
        status.state = WorkerState::Idle;
        status.progress = 0;
        status.message.clear();
        status.items_done = 0;
        status.items_total = 0;
        let status = status.clone();
        Self::broadcast(&mut state, now, name, status);
    }

    pub fn status(&self, name: &str) -> Option<WorkerStatus> {
        self.state().statuses.get(name).cloned()
    }

    pub fn snapshot(&self) -> Vec<(String, WorkerStatus)> {
        self.state()
            .statuses
            .iter()
            .map(|(name, status)| (name.clone(), status.clone()))
            .collect()
    }

    pub fn subscribe(&self) -> mpsc::Receiver<WorkerStatusUpdate> {
        self.subscribe_mapped(Clone::clone)
    }

    pub fn subscribe_mapped<T, F>(&self, map: F) -> mpsc::Receiver<T>
    where
        T: Send + 'static,
        F: Fn(&WorkerStatusUpdate) -> T + Send + Sync + 'static,
    {
        let (sender, receiver) = mpsc::channel(SUBSCRIBER_CAPACITY);
        let closed_sender = sender.clone();
        let id = self.next_subscriber_id.fetch_add(1, Ordering::Relaxed);
        let mut state = self.state();
        state
            .subscribers
            .retain(|_, subscriber| !(subscriber.is_closed)());
        state.subscribers.insert(
            id,
            Subscriber {
                send: Box::new(move |update| match sender.try_send(map(update)) {
                    Ok(()) | Err(mpsc::error::TrySendError::Full(_)) => SubscriberState::Active,
                    Err(mpsc::error::TrySendError::Closed(_)) => SubscriberState::Closed,
                }),
                is_closed: Box::new(move || closed_sender.is_closed()),
            },
        );
        receiver
    }

    pub fn subscriber_count(&self) -> usize {
        let mut state = self.state();
        state
            .subscribers
            .retain(|_, subscriber| !(subscriber.is_closed)());
        state.subscribers.len()
    }

    fn broadcast(
        state: &mut StatusState,
        timestamp: DateTime<Utc>,
        name: &str,
        status: WorkerStatus,
    ) {
        let update = WorkerStatusUpdate {
            worker_name: name.to_owned(),
            status,
            timestamp,
        };
        state
            .subscribers
            .retain(|_, subscriber| (subscriber.send)(&update) == SubscriberState::Active);
    }

    fn state(&self) -> MutexGuard<'_, StatusState> {
        self.state
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner)
    }
}
